File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case14_tableAndStreamAndSql/Case11_TableApiState.java
package com.yw.flink.example.javacases.case14_tableAndStreamAndSql;

import com.yw.flink.example.StationLog;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;

/**
 * Flink Table API 状态保存测试
 */
public class Case11_TableApiState {
    public static void main(String[] args) {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env);

        tableEnv.getConfig().set("table.exec.state.ttl", "5000");

        //创建DataStream
        SingleOutputStreamOperator<StationLog> stationLogDs = env.socketTextStream("node5", 9999)
                .map((MapFunction<String, StationLog>) line -> {
                    String[] arr = line.split(",");
                    return new StationLog(arr[0], arr[1], arr[2], arr[3], Long.valueOf(arr[4]), Long.valueOf(arr[5]));
                });

        //将DataStream转换成Table
        tableEnv.createTemporaryView("station_log_tbl", stationLogDs);

        //通过SQL统计每个基站的通话时长
        Table result = tableEnv.sqlQuery("select sid,sum(duration) as total_duration from station_log_tbl group by sid");
        result.execute().print();

    }

}

File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case14_tableAndStreamAndSql/Case11_TableApiState.scala
package com.yw.flink.example.scalacases.case14_tableAndStreamAndSql

import com.yw.flink.example.StationLog
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.table.api.Table
import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment

/**
  * Flink Table API 状态保存测试
  */
object Case11_TableApiState {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    import org.apache.flink.streaming.api.scala._

    val tableEnv: StreamTableEnvironment = StreamTableEnvironment.create(env)
    tableEnv.getConfig.set("table.exec.state.ttl", "5000")
    val ds: DataStream[StationLog] = env.socketTextStream("node5", 9999)
      .map(line => {
        val arr: Array[String] = line.split(",")
        StationLog(arr(0), arr(1), arr(2), arr(3), arr(4).toLong, arr(5).toLong)
      })

    tableEnv.createTemporaryView("tbl", ds)

    val result: Table = tableEnv.sqlQuery("select sid,sum(duration) as total from tbl group by sid")
    result.execute().print()
  }

}
